### PR TITLE
[#3564] pam_password: CSP redirect only enables SSL if necessary (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses lingering issue found here (i.e. not the main issue, which has already been fixed): https://github.com/irods/irods/issues/3564#issuecomment-1256617949

Confirmed that the issue is resolved only at the bench because the issue manifests as log noise, as far as I can tell. If anybody can think of a way in which this might present a more serious problem which we can capture in a test, I will write an automated test.

### CODE REVIEW

File: plugins/auth/src/pam_password.cpp [hunk 1]
```diff
@@ -271,11 +271,26 @@ namespace irods
...
-                if (const auto ec = sslStart(host->conn); ec) {
-                    THROW(ec, "could not establish SSL connection");
-                }
+                // Need to enable SSL here if it is not already being used because the PAM password
+                // is forwarded to the provider in the clear.
+                // clang-format off
+                const bool using_ssl = 0 == std::strncmp(
+                        irods::CS_NEG_USE_SSL.c_str(),
+                        host->conn->negotiation_results, // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+                        MAX_NAME_LEN);
```
Comment (korydraughn) [inline plugins/auth/src/pam_password.cpp:280]: Indentation?

**Review by korydraughn — COMMENTED** [review]
Looks good.

File: plugins/auth/src/pam_password.cpp (hunk 1 above)
Comment (korydraughn) [reply plugins/auth/src/pam_password.cpp:280]: I was referring to the indentation appearing to use more than 4 spaces.
```cpp
const bool using_ssl = 0 == std::strncmp(
￼        irods::CS_NEG_USE_SSL.c_str(),
￼        host->conn->negotiation_results, // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
￼        MAX_NAME_LEN);
```
VS
```cpp
const bool using_ssl = 0 == std::strncmp(
￼   irods::CS_NEG_USE_SSL.c_str(),
￼   host->conn->negotiation_results, // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
￼   MAX_NAME_LEN);
```
The `NOLINTBEGIN`/`END` thing is not available in Clang 13's clang-tidy. I've been using those knowing that we'll be upgrading the compiler some time in the future.

Comment (alanking) [reply plugins/auth/src/pam_password.cpp:280]: Ah, okay. Thanks. Will fix this up.

**Review by korydraughn — APPROVED** [review]
If all tests pass, add the pounds and merge.

**Comment by trel** [comment]
agreed, remember to squash first

**Comment by alanking** [comment]
Well, the only test which would exercise this code is skipped when run in topology from the resource server or when SSL is enabled, so there's that. It made me think that maybe we should consider making a mode for the test suite which uses pam_password. The rest of the test suite outside of `test_auth` uses native authentication and would never hit this.

Anyway, I confirmed at the bench that the fix addresses the linked issue and the test suite is passing.

**Comment by korydraughn** [comment]
> It made me think that maybe we should consider making a mode for the test suite which uses pam_password.

Yes. We need to exercise as much of the code as possible. Try thinking about what a pam_password environment would look like. Perhaps it can be captured in an issue if not already.

> Anyway, I confirmed at the bench that the fix addresses the linked issue and the test suite is passing.

Good. Please merge when ready.

**Review by trel — APPROVED** [review]

**Comment by alanking** [comment]
I created https://github.com/irods/irods/issues/6785 for the authentication method specification when running tests. Thanks!

**Pull request merged**